Packages/com.unity.inputsystem/InputSystem/Plugins/XInput/XboxOneGamepad.cs

```diff
@@ -161,6 +208,76 @@ protected override void FinishSetup(InputDeviceBuilder builder)
             paddle2 = builder.GetControl<ButtonControl>(this, "paddle2");
             paddle3 = builder.GetControl<ButtonControl>(this, "paddle3");
             paddle4 = builder.GetControl<ButtonControl>(this, "paddle4");
+        }
+
+        public ulong gamepadId
+        {
+            get
+            {
+                UpdatePadSettings();
+                return m_GamepadId;
+            }
+        }
+
+        public int xboxOneUserId
+        {
+            get
+            {
+                UpdatePadSettings();
+                return m_XboxOneUserId;
+            }
+        }
+
+        public new static List<XboxOneGamepad> all
+        {
+            // Needs to return a copy just in case anything else might be enumerating the returned list while gamepads are being added and removed.
+            // This will generate garbage and should be used sparingly.
+            get { return new List<XboxOneGamepad>(s_Devices); }
+        }
+
+        public static XboxOneGamepad GetByGamepadId(ulong gamepadId)
+        {
+            for (int i = 0; i < s_Devices.Count; i++)
+            {
+                if (s_Devices[i] != null && s_Devices[i].gamepadId == gamepadId)
+                {
+                    return s_Devices[i];
+                }
+            }
+
+            return null;
+        }
+
+        protected override void OnAdded()
+        {
+            base.OnAdded();
+
+            s_Devices.Add(this);
+        }
+
+        protected override void OnRemoved()
+        {
+            base.OnRemoved();
+
+            for (int i = 0; i < s_Devices.Count; i++)
+            {
+                if (s_Devices[i] != null && s_Devices[i].gamepadId == gamepadId)
+                {
+                    s_Devices.RemoveAt(i);

```

Comment:
With `OnAdded` doing `Add(this)`, seems like this should just be `Remove(this)`. When switched to array, use ArrayHelpers.